### PR TITLE
Cosmetic changes and fix to harpoon bolts and rockets

### PR DESCRIPTION
**Pull request by Kubsy**
branch by Gancian, it is finished and should be ready for pull request

- fixed the harpoon bug when shooting vertically
- added specific constants for harpoon and for arrows, so that they can be later refractored for lua costumization
- added splash for rockets when entering water
- added sound effect when shooting harpoons from the UPV

### CODE REVIEW

File: TombEngine/Game/Lara/lara_one_gun.cpp [hunk 1]
```diff
@@ -185,7 +190,8 @@ void AnimateShotgun(ItemInfo& laraItem, LaraWeaponType weaponType)
 						break;
 
 					case LaraWeaponType::HarpoonGun:
-						FireHarpoon(laraItem);
+					
+						FireHarpoon(laraItem, nullptr);
```
Comment (Lwmte) [inline TombEngine/Game/Lara/lara_one_gun.cpp:194]: If you add new argument for the function which is mostly always passed as `nullptr`, you can make this argument default in declaration (.h file).

File: TombEngine/Game/Lara/lara_one_gun.cpp [hunk 2]
```diff
@@ -1459,7 +1467,7 @@ void HandleProjectile(ItemInfo& projectile, ItemInfo& emitter, const Vector3i& p
 	bool hasHit = false;
 	bool hasHitNotByEmitter = false;
 	bool isExplosive = type >= ProjectileType::Explosive;
-	bool isShatterable = type != ProjectileType::Harpoon;
+	bool isShatterable = true;
```
Comment (Lwmte) [inline TombEngine/Game/Lara/lara_one_gun.cpp:1470]: This is incorrect. Why do you remove this condition? Harpoon should remain stuck into the room geometry until it shatters by itself after some interval.

File: TombEngine/Objects/TR3/Vehicles/upv.cpp [hunk 3]
```diff
@@ -223,17 +223,24 @@ namespace TEN::Entities::Vehicles
 
 	static void FireUPVHarpoon(ItemInfo* UPVItem, ItemInfo* laraItem)
 	{
-		auto* harpoon = FireHarpoon(*laraItem);
-
-		if (harpoon == nullptr)
-			return;
+		// FireHarpoon(*laraItem, nullptr);
```
Comment (Lwmte) [inline TombEngine/Objects/TR3/Vehicles/upv.cpp:226]: According to coding conventions, you should not leave randomly commented code lines in the codebase. If you are not sure if your new code is stable, more tests should be made.

File: TombEngine/Objects/TR3/Vehicles/upv.cpp [hunk 4]
```diff
@@ -223,17 +223,24 @@ namespace TEN::Entities::Vehicles
...
+		// GianC.- Use this if the code misbehaves. It shoots the harpoon from pos 0 (under Lara)
 
 		auto UPV = GetUPVInfo(UPVItem);
-		harpoon->Pose.Position = GetJointPosition(UPVItem, UPV_JOINT_TURBINE, Vector3i((UPV->HarpoonLeft ? 22 : -22), 24, 230));
-		harpoon->Pose.Orientation = EulerAngles(UPVItem->Pose.Orientation.x, UPVItem->Pose.Orientation.y, 0);
+		//auto orient = EulerAngles(UPVItem->Pose.Orientation.x, UPVItem->Pose.Orientation.y, 0);
```
Comment (Lwmte) [inline TombEngine/Objects/TR3/Vehicles/upv.cpp:230]: As above. If this code is obsolete, remove it.

File: TombEngine/Objects/TR3/Vehicles/upv.cpp [hunk 5]
```diff
@@ -223,17 +223,24 @@ namespace TEN::Entities::Vehicles
...
+
+		if (UPV->Flags & UPV_FLAG_SURFACE)
+		SoundEffect(SFX_TR4_HARPOON_FIRE_DRY, &harpoonPose, SoundEnvironment::Always);
+		
+		else
+		SoundEffect(SFX_TR4_HARPOON_FIRE_UNDERWATER, &harpoonPose, SoundEnvironment::Always);
 
-		harpoon->ItemFlags[0] = 1;
-		UPV->HarpoonLeft = !UPV->HarpoonLeft;	
+		//harpoon->ItemFlags[0] = 1;
```
Comment (Lwmte) [inline TombEngine/Objects/TR3/Vehicles/upv.cpp:241]: As above.

**Review by Lwmte — CHANGES_REQUESTED** [review]

File: TombEngine/Game/Lara/lara_one_gun.cpp [hunk 6]
```diff
@@ -556,109 +562,99 @@ void UndrawShotgunMeshes(ItemInfo& laraItem, LaraWeaponType weaponType)
 	}
 }
 
-ItemInfo* FireHarpoon(ItemInfo& laraItem)
+void FireHarpoon(ItemInfo& laraItem, Pose* pos)
```
Comment (Sezzary) [inline TombEngine/Game/Lara/lara_one_gun.cpp:565]: I would recommend using `std::optional<Pose>` here instead of a `Pose*` pointer.

**Comment by Kubsy** [comment]
Harpoons shatter upon contact with the wall, which is a regression and should be fixed.

File: TombEngine/Game/Lara/lara_one_gun.cpp [hunk 7]
```diff
@@ -556,109 +561,98 @@ void UndrawShotgunMeshes(ItemInfo& laraItem, LaraWeaponType weaponType)
...
+			harpoonItem.Pose.Orientation += player.ExtraTorsoRot;
+	}
+
+	harpoonItem.Animation.Velocity.z = HARPOON_VELOCITY;
+	harpoonItem.HitPoints = HARPOON_TIME;
 
-	Rumble(0.2f, 0.1f);
 	AddActiveItem(itemNumber);
 
+	harpoonItem.ItemFlags[0] = (int)ProjectileType::Normal;
```
Comment (Lwmte) [inline TombEngine/Game/Lara/lara_one_gun.cpp:629]: Why?! Harpoon projectile type is harpoon!

File: TombEngine/Game/Lara/lara_one_gun.cpp [hunk 8]
```diff
@@ -556,109 +561,98 @@ void UndrawShotgunMeshes(ItemInfo& laraItem, LaraWeaponType weaponType)
...
-
+	
 void HarpoonBoltControl(short itemNumber)
 {
 	auto& harpoonItem = g_Level.Items[itemNumber];
 
-	if (harpoonItem.HitPoints < HARPOON_TIME)
+	if (TestEnvironment(ENV_FLAG_WATER, &harpoonItem))
 	{
-		if (harpoonItem.HitPoints > 0)
```
Comment (Lwmte) [inline TombEngine/Game/Lara/lara_one_gun.cpp:626]: Where this code have gone? It shattered harpoon after a time interval when it stucked into room geometry!

**Review by Lwmte — CHANGES_REQUESTED** [review]
This pull request introduces serious regressions which have to be fixed.

**Comment by Lwmte** [comment]
Clipboard-1

Also horizontal firing angle was not fixed (when you look sideways underwater).

**Comment by Kubsy** [comment]
UPV harpoon whoot sfx plays even if there’s amy harpoons left

**Pull request merged**